cpp/tests/copying/scatter_to_tables_tests.cu

```diff
@@ -0,0 +1,330 @@
+    for(gdf_size_type table_n_rows : {0, 1})
+      for(gdf_size_type scatter_size : {0, 1}) {
+        //std::cout << table_n_cols << table_n_rows << scatter_size << std::endl;
+
+        //data
+        std::vector<cudf::test::column_wrapper<TypeParam>> v_colwrap(
+            table_n_cols, {table_n_rows,
+              [](gdf_index_type row) { return static_cast<TypeParam>(row); },
+              [](gdf_index_type row) { return false; }});
+        std::vector<gdf_column *> v_cols(table_n_cols);
+        for (size_t i = 0; i < v_colwrap.size(); i++)
+          v_cols[i] = v_colwrap[i].get();
+
+        //input datatypes
+        cudf::table input_table{v_cols};
+        cudf::test::column_wrapper<gdf_index_type> scatter_map(scatter_size,
+          [](gdf_index_type row) { return row; }, false);
+        std::vector<cudf::table> output_tables;
+        if (table_n_cols == 0) table_n_rows=0;
+        if(scatter_size != table_n_rows) {
+          CUDF_EXPECT_THROW_MESSAGE(
+            output_tables = cudf::scatter_to_tables(input_table, scatter_map),
+            "scatter_map length is not equal to number of rows in input table.");
+        } else {
+          EXPECT_NO_THROW(
+            output_tables = cudf::scatter_to_tables(input_table, scatter_map));
+            EXPECT_EQ(size_t(scatter_size), output_tables.size());
+        }
+      }
+}
+
+template <typename T>
+auto scatter_columns(
+    std::vector<std::vector<T>> input_cols_data,
+    std::vector<std::vector<gdf_valid_type>> input_cols_bitmask,
+    std::vector<gdf_index_type> scatter_map)
+{
+  auto max = *std::max_element(scatter_map.begin(), scatter_map.end());
+  std::vector<std::vector<std::vector<T>>> output_cols_data(max+1);
+  std::vector<std::vector<std::vector<gdf_valid_type>>> output_cols_bitmask(max+1);
+  std::vector<std::vector<gdf_size_type>> output_cols_null_count(max+1);
+  auto num_cols = input_cols_data.size();
+  for (auto g = 0; g <= max; g++)
+  {
+    auto n_output = std::count(scatter_map.begin(), scatter_map.end(), g);
+    for (size_t c = 0; c < num_cols; c++)
+    {
+      std::vector<T> output_col_data(n_output);
+      std::vector<gdf_valid_type> output_col_bitmask(gdf_valid_allocation_size(n_output), ~gdf_valid_type(0));
+      gdf_index_type j = 0, nullc=0;
+      for (size_t r=0; r< scatter_map.size(); r++)
+      {
+        if (g == scatter_map[r])
+        {
+          output_col_data[j]=input_cols_data[c][r];
+          if(!gdf_is_valid(input_cols_bitmask[c].data(), r))
+          {
+            bit_mask::clear_bit_unsafe(
+              reinterpret_cast<bit_mask::bit_mask_t*>(output_col_bitmask.data()), j);
+            nullc++;
+          }
+          j++;
+        }
+      }
+      output_cols_data[g].push_back(output_col_data);
+      output_cols_bitmask[g].push_back(output_col_bitmask);
+      output_cols_null_count[g].push_back(nullc);
+      assert(n_output == j);
+    }
+    assert(num_cols == output_cols_data[g].size());
```

Comment:
I used assert because this is CPU reference solution function. I assumed `ASSERT_EQ` or `EXPECT_*` is to test the libcudf library functions outputs only.

Reply:
But assert() only tests the CPU results in debug builds. ASSERT is not limited to GPU results, gtest is a generic unit testing framework.

Reply:
I understand that. By convention, all tests are meant to be testing _libcudf_ library. so, i refrained from adding gtest assert statements for reference calculations. I have added these now.